cli-plugins/manager/hooks.go

```diff
@@ -0,0 +1,140 @@
+// is invoked.
+type HookMessage struct {
+	// TODO: should we have an enum to typify the kind of hook
+	// (such as `Next Steps`, `Warning`, etc.)?
+	Template string
+}
+
+// TemplateReplaceSubcommandName returns a hook template string
+// that will be replaced by the CLI subcommand being executed
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran the subcommand: ` + TemplateReplaceSubcommandName(),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran the subcommand: run`
+func TemplateReplaceSubcommandName() string {
+	return hookTemplateCommandName
+}
+
+// TemplateReplaceFlagValue returns a hook template string
+// that will be replaced by the flags value.
+//
+// Example:
+//
+//	h := HookMessage{
+//	  Template: `you ran a container named: ` + TemplateReplaceFlagValue("name"),
+//	}
+//
+// when being executed after the command:
+// `docker run --name "my-container" alpine`
+// will result in the message:
+// `you ran a container named: my-container`
+func TemplateReplaceFlagValue(flag string) string {
+	return fmt.Sprintf(hookTemplateFlagValue, flag)
+}
+
+// RunPluginHooks calls the hook subcommand for all present
+// CLI plugins that declare support for hooks in their metadata
+// and parses/prints their responses
+func RunPluginHooks(dockerCli command.Cli, cmd *cobra.Command, args []string) error {
+	plugins, err := ListPlugins(dockerCli, cmd)
+	if err != nil {
+		return err
+	}
+	subCommand, _, err := cmd.Find(args)
+	if err != nil {
+		return err
+	}
+
+	hookMessages := make([]string, 0, len(plugins))
+	for _, p := range plugins {
+		p := p
+		if !p.supportsHookCommand(subCommand.Name()) {
+			continue
+		}
+
+		hookReturn, err := p.RunHook(subCommand.Name())
+		// Skip misbehaving plugins, but don't halt execution
+		if err != nil {
+			continue
+		}
+		hookMessages = append(hookMessages, parseHooks(hookReturn, subCommand))
+	}
+
+	printHookMessages(dockerCli, hookMessages)
+	return nil
+}
+
+const (
+	hookTemplateCommandName = "{docker-cli-plugin-hook-cmd-name}"
+	hookTemplateFlagValue   = "{docker-cli-plugin-hook-flag-%s}"
+)
+
+func parseHooks(hookMsg HookMessage, cmd *cobra.Command) string {
+	ret := hookMsg.Template
+	for {
+		parsed, foundMatch := parseFlagValue(ret, cmd)
+		ret = parsed
+		if !foundMatch {
+			break
+		}
+	}
+	return strings.ReplaceAll(ret, hookTemplateCommandName, cmd.Name())
+}
+
+var flagRegex = regexp.MustCompile(fmt.Sprintf(hookTemplateFlagValue, `([\w-]+)`))
+
+func parseFlagValue(template string, cmd *cobra.Command) (string, bool) {
+	matches := flagRegex.FindStringSubmatch(template)
+	if len(matches) == 0 {
+		return template, false
+	}
+	for _, m := range matches[1:] {
+		flag := cmd.Flag(m)
+		if flag == nil {
+			continue
+		}
+		value := flag.Value.String()
+		template = strings.ReplaceAll(template, TemplateReplaceFlagValue(m), value)
```

Comment:
Should we use an actual Go template for this? (it just occurred to me that we are doing a DYI variant of that here?)

Reply:
When implementing this my thought was just to provide some simple, very easy to use API that plugins could call into to build their templates. We can definitely replace that with actual Go templates, though I do worry it might result in more complex implementation and, in that case, we should try to figure out how to best expose/make clear to Hooks API users what information they can/cannot access with Go templates.

Another option is to keep the current API and just use Go templates under the hood for implementation, which might make sense (but I'd need to dig into it a bit more).

Reply:
@thaJeztah switched to using Go templates! :)